diff --git a/config.def.h b/config.def.h
index 618d4d8..5abc827 100644
--- a/config.def.h
+++ b/config.def.h
@@ -24,3 +24,8 @@ static unsigned int lines      = 0;
  * for example: " /?\"&[]"
  */
 static const char worddelimiters[] = " ";
+
+/* Size of the window border */
+static unsigned int border_width = 2;
+/* If set to 0, use --border to get above specified border width, set to 1 to not have tp use --border parameter everytime */
+static unsigned int border = 0;
diff --git a/config.h b/config.h
index 618d4d8..5abc827 100644
--- a/config.h
+++ b/config.h
@@ -24,3 +24,8 @@ static unsigned int lines      = 0;
  * for example: " /?\"&[]"
  */
 static const char worddelimiters[] = " ";
+
+/* Size of the window border */
+static unsigned int border_width = 2;
+/* If set to 0, use --border to get above specified border width, set to 1 to not have tp use --border parameter everytime */
+static unsigned int border = 0;
diff --git a/dmenu b/dmenu
index 07eb9aa..b1675fc 100755
Binary files a/dmenu and b/dmenu differ
diff --git a/dmenu.c b/dmenu.c
index febf5db..7503b91 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -876,9 +876,16 @@ setup(void)
 	swa.background_pixel = scheme[SchemeNorm][ColBg].pixel;
 	swa.event_mask = ExposureMask | KeyPressMask | VisibilityChangeMask |
 	                 ButtonPressMask | PointerMotionMask;
-	win = XCreateWindow(dpy, parentwin, x, y, mw, mh, 0,
-	                    CopyFromParent, CopyFromParent, CopyFromParent,
-	                    CWOverrideRedirect | CWBackPixel | CWEventMask, &swa);
+    if(border){
+	    win = XCreateWindow(dpy, parentwin, x, y, mw, mh, border_width,
+	                        CopyFromParent, CopyFromParent, CopyFromParent,
+	                        CWOverrideRedirect | CWBackPixel | CWEventMask, &swa);
+	    XSetWindowBorder(dpy, win, scheme[SchemeSel][ColBg].pixel);
+    }else{
+        win = XCreateWindow(dpy, parentwin, x, y, mw, mh, 0,
+	                        CopyFromParent, CopyFromParent, CopyFromParent,
+	                        CWOverrideRedirect | CWBackPixel | CWEventMask, &swa);
+    }	XSetClassHint(dpy, win, &ch);
 	XSetClassHint(dpy, win, &ch);
 
 
@@ -963,7 +970,10 @@ main(int argc, char *argv[])
 			fast = 1;
 		else if (!strcmp(argv[i], "-c"))   /* centers dmenu on screen */
 			centered = 1;
-		else if (!strcmp(argv[i], "-i")) { /* case-insensitive item matching */
+        else if (!strcmp(argv[i], "--border")){   /* show border around dmenu */
+            border = 1;
+			/*border_width = atoi(argv[++i]);*/
+        }else if (!strcmp(argv[i], "-i")) { /* case-insensitive item matching */
 			fstrncmp = strncasecmp;
 			fstrstr = cistrstr;
 		} else if (i + 1 == argc)
diff --git a/dmenu.o b/dmenu.o
index 7e002fd..9753270 100644
Binary files a/dmenu.o and b/dmenu.o differ
