diff --git a/README.md b/README.md
index 01078e2..b1b6aaa 100644
--- a/README.md
+++ b/README.md
@@ -136,6 +136,7 @@ Super key is an alternative name for the Windows key or Command key when using L
 - pertag
 - placemouse
 - preserveonrestart
+- removeborder
 - resizepoint
 - restartsig
 - setborderpx
diff --git a/config.def.h b/config.def.h
index 026c355..00a5fe8 100644
--- a/config.def.h
+++ b/config.def.h
@@ -25,6 +25,8 @@ static const Bool viewontag         = True;     /* Switch view on tag switch */
 static const int swallowfloating    = 0;        /* 1 means swallow floating windows by default */
 static const unsigned int baralpha = 0xd0;
 static const unsigned int borderalpha = OPAQUE;
+static const unsigned int rmborder = 1; /* Set to 1 to remove border when only one window is visible on screen */
+
 static char normbgcolor[]           = "#222222";
 static char normbordercolor[]       = "#444444";
 static char normfgcolor[]           = "#bbbbbb";
diff --git a/dwm b/dwm
index ee7cd09..f9d977f 100755
Binary files a/dwm and b/dwm differ
diff --git a/dwm.c b/dwm.c
index 5f788ae..5b62b1d 100644
--- a/dwm.c
+++ b/dwm.c
@@ -2397,6 +2397,8 @@ void
 resizeclient(Client *c, int x, int y, int w, int h)
 {
 	XWindowChanges wc;
+	unsigned int n;
+	Client *nbc;
 
 	c->oldx = c->x; c->x = wc.x = x;
 	c->oldy = c->y; c->y = wc.y = y;
@@ -2407,6 +2409,20 @@ resizeclient(Client *c, int x, int y, int w, int h)
 		return;
 
 	wc.border_width = c->bw;
+    
+    if(rmborder){
+        for (n = 0, nbc = nexttiled(selmon->clients); nbc; nbc = nexttiled(nbc->next), n++);
+
+	    if (c->isfloating || selmon->lt[selmon->sellt]->arrange == NULL) {
+	    } else {
+		    if (selmon->lt[selmon->sellt]->arrange == monocle || n == 1) {
+			wc.border_width = 0;
+			c->w = wc.width += c->bw * 2;
+			c->h = wc.height += c->bw * 2;
+		    }
+	    }
+    }
+
 	XConfigureWindow(dpy, c->win, CWX|CWY|CWWidth|CWHeight|CWBorderWidth, &wc);
 	configure(c);
 	XSync(dpy, False);
diff --git a/dwm.o b/dwm.o
index ca2dae3..a66fbfa 100644
Binary files a/dwm.o and b/dwm.o differ
